crates/matrix-sdk/src/client/builder.rs

```diff
@@ -371,63 +402,88 @@ impl ClientBuilder {
+                // When discovery fails, or the input isn't a valid server name, fallback to
+                // trying a homeserver URL if supplied.
+                if let Ok(homeserver_url) = Url::parse(&server_name_or_url) {
+                    // Make sure the URL is definitely for a homeserver.
+                    if http_client
+                        .send(
+                            get_supported_versions::Request::new(),
+                            Some(RequestConfig::short_retry()),
+                            homeserver_url.to_string(),
+                            None,
+                            &[MatrixVersion::V1_0],
+                            Default::default(),
+                        )
+                        .await
+                        .is_ok()
+                    {
+                        result = Ok((homeserver_url.to_string(), None));
+                    } else {
+                        // If it isn't, we'll throw the discovery error instead.
+                    }
                 }
 
-                url
+                result?
             }
-            HomeserverConfig::ServerName { server: server_name, protocol } => {
-                debug!("Trying to discover the homeserver");
-
-                let homeserver = match protocol {
-                    UrlScheme::Http => format!("http://{server_name}"),
-                    UrlScheme::Https => format!("https://{server_name}"),
-                };
-
-                let well_known = http_client
-                    .send(
-                        discover_homeserver::Request::new(),
-                        Some(RequestConfig::short_retry()),
-                        homeserver,
-                        None,
-                        &[MatrixVersion::V1_0],
-                        Default::default(),
-                    )
-                    .await
-                    .map_err(|e| match e {
-                        HttpError::Api(err) => ClientBuildError::AutoDiscovery(err),
-                        err => ClientBuildError::Http(err),
-                    })?;
-
-                #[cfg(feature = "experimental-oidc")]
-                {
-                    authentication_server_info = well_known.authentication;
-                    allow_insecure_oidc = matches!(protocol, UrlScheme::Http);
-                }
-
-                #[cfg(feature = "experimental-sliding-sync")]
-                if let Some(proxy) = well_known.sliding_sync_proxy.map(|p| p.url) {
-                    sliding_sync_proxy = Url::parse(&proxy).ok();
-                }
+        };
 
-                debug!(
-                    homeserver_url = well_known.homeserver.base_url,
-                    "Discovered the homeserver"
-                );
+        #[allow(unused_variables)]
+        if let Some(well_known) = well_known {
+            #[cfg(feature = "experimental-oidc")]
+            {
+                authentication_server_info = well_known.authentication;
```

Comment:
I think that now, both `authentication_server_info` and `allow_insecure_url` could be defined only around here? (instead of before the match on the `HomeserverConfig`, making the definitions closer to their uses)